### PR TITLE
Sequences & Maps

### PR DESCRIPTION
**Pull request by JM4ier**
additional functions for the two mathematical data types and support for the Index trait (WIP)

### CODE REVIEW

**Comment by JM4ier** [comment]
I can't reproduce the failing test, instead I get the following error when running the test:

```
./x.py verify-test prusti-tests/tests/verify_overflow/fail/core_proof/arrays.rs
```
```
error: [Prusti internal error] Prusti encountered an unexpected internal error
  |
  = help: This could be caused by too small assertion timeout. Try increasing it by setting the configuration parameter ASSERT_TIMEOUT to a larger value.
  = note: We would appreciate a bug report: https://github.com/viperproject/prusti-dev/issues/new
  = note: Details: unregistered verification error: [predicate.not.wellformed:receiver.not.injective] Predicate might not be well-formed. Receiver of OwnedNonAliased$I32(index_place$$array$I32(place, constructor$Snap$Usize$(index)), root_address, snapshot[index]) [<no position>]  might not be injective. (<no position>)
```

File: prusti-viper/src/encoder/middle/core_proof/adts/interface.rs [hunk 1]
```diff
@@ -114,10 +114,16 @@ pub(in super::super) trait AdtsInterface {
     /// and its result is used to guard the injectivity property. This is
     /// intended to be used by the snapshot encoder to supply call to the
     /// validity function.
+    ///
+    /// If `generate_top_down_injectivity_axiom` is true, do not generate
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/adts/interface.rs]: Shouldn't it be:
```suggestion
    /// If `use_main_constructor_destructors` is true, do not generate
```

File: prusti-viper/src/encoder/middle/core_proof/snapshots/into_snapshot/common/mod.rs [hunk 2]
```diff
@@ -455,13 +456,30 @@ pub(super) trait IntoSnapshotLowerer<'p, 'v: 'p, 'tcx: 'v> {
             .collect::<Result<Vec<_>, _>>()?;
         let mut args =
             self.expression_vec_to_snapshot(lowerer, &app.arguments, expect_math_bool)?;
+        log::debug!(
+            "BFASNAPSHOT  ty_args:{ty_args:?} app_args:{:?} args:{args:?}",
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/snapshots/into_snapshot/common/mod.rs]: What does `BFASNAPSHOT` mean?

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 3]
```diff
@@ -1290,237 +1290,217 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         cleanup: &Option<mir::BasicBlock>,
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: This function looks very large. Could you please move the builtin stuff to a separate file? (Similarly to `lifetimes.rs`.)

File: prusti-viper/src/encoder/mir/pure/interpreter/mod.rs [hunk 4]
```diff
@@ -626,15 +680,33 @@ impl<'p, 'v: 'p, 'tcx: 'v> ExpressionBackwardInterpreter<'p, 'v, 'tcx> {
...
+                    }) => {
+                        let ref_type = encoded_lhs.get_type().clone();
+                        builtin((LookupMap, ref_type))
+                    }
+                    Type::Reference(Reference {
+                        target_type: box Type::Sequence(_),
+                        ..
+                    }) => {
+                        let ref_type = encoded_lhs.get_type().clone();
+                        //assert!(encoded_args[0].get_type().is_sequence(), "Expected sequence type, got {:?}", encoded_args[0].get_type());
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/pure/interpreter/mod.rs]:
```suggestion
```

**Review by vakaras — CHANGES_REQUESTED** [review]
Looks good. I have left some minor comments.

**Comment by vakaras** [comment]
> I can't reproduce the failing test, instead I get the following error when running the test:
> 
> ```
> ./x.py verify-test prusti-tests/tests/verify_overflow/fail/core_proof/arrays.rs
> ```
> 
> ```
> error: [Prusti internal error] Prusti encountered an unexpected internal error
>   |
>   = help: This could be caused by too small assertion timeout. Try increasing it by setting the configuration parameter ASSERT_TIMEOUT to a larger value.
>   = note: We would appreciate a bug report: https://github.com/viperproject/prusti-dev/issues/new
>   = note: Details: unregistered verification error: [predicate.not.wellformed:receiver.not.injective] Predicate might not be well-formed. Receiver of OwnedNonAliased$I32(index_place$$array$I32(place, constructor$Snap$Usize$(index)), root_address, snapshot[index]) [<no position>]  might not be injective. (<no position>)
> ```

It seems that we have a flaky test. Please ignore it, I will fix it.

File: prusti-viper/src/encoder/middle/core_proof/adts/interface.rs (hunk 1 above)
Comment (JM4ier) [reply prusti-viper/src/encoder/middle/core_proof/adts/interface.rs]: I think that must've been one of your commits that got mixed in when I squashed the commits.

File: prusti-viper/src/encoder/middle/core_proof/snapshots/into_snapshot/common/mod.rs (hunk 2 above)
Comment (JM4ier) [reply prusti-viper/src/encoder/middle/core_proof/snapshots/into_snapshot/common/mod.rs]: It was a temporary debug statement to find out where type and expression inconsistencies are introduced.
I removed it.

**Review by vakaras — APPROVED** [review]

**Pull request merged**